Add RowNumber task to the concurrent test in SharedArbitrationTest

duanmeng opened the pull request:
PR #7846 fixed the spill twice issue in `RowNumber`, it is nice to add the
`RowNumber` task back to the concurrent test in SharedArbitrationTest.

xiaoxmeng reviewed (approved):
@duanmeng LGTM. Thanks for the fix. Is it possible to have a dedicated test case for this? Thanks!

mbasmanova commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -14,19 +14,76 @@
+    executor_ = std::make_unique<folly::CPUThreadPoolExecutor>(32);
   }
+
+  void setupMemory(
```
I expect this logic will be repeated in many tests. Can we extract it into helper methods in the base class?

duanmeng replied:
Got it, I will fix this once the refactor PR created by @xiaoxmeng is merged.

mbasmanova commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -218,4 +275,40 @@ TEST_F(RowNumberTest, largeInput) {
+      makeFlatVector<int64_t>({1, 2, 3, 4, 5, 6, 7}),
+  });
+
+  createDuckDbTable({data});
```
This statement is not needed since this test is not expected to produce results that need comparing.

mbasmanova commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -218,4 +275,40 @@ TEST_F(RowNumberTest, largeInput) {
+  createDuckDbTable({data});
+  setupMemory(1 * MB);
+  try {
+    AssertQueryBuilder(duckDbQueryRunner_)
```
remove duckDbQueryRunner_ and replace assertResults with copyResults(pool())

mbasmanova commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -218,4 +275,40 @@ TEST_F(RowNumberTest, largeInput) {
+            "FROM ("
+            " SELECT *, row_number() over (partition by c0) AS rn FROM tmp) "
+            "WHERE rn <= 2");
+  } catch (const VeloxException& e) {
```
Use VELOX_ASSERT_THROW and if needed add a similar macro for memory-specific exceptions. This pattern is buggy because the test allows the query to succeed. You would need to add VELOX_FAIL in the 'try' body to fix that, but people always forget.

duanmeng replied:
@mbasmanova Yes, this version is buggy and I found it is not guaranteed to trigger the 'hasSpilled' check logic added in this PR.

I've updated this test (**`RowNumberTest.reclaimFail`**), it is designed to make `RowNumber::maybeReserve` spill through `RowNumber::reclaim` and return false to guard the 'hasSpilled' check logic added in this PR. It creates multiple tasks to trigger the memory reclaim, some tasks may fail with memory-related exceptions, which is acceptable and some may not, so I catch the expected exception with the expected error code and rethrow the unexpected exception. But it is kind of complicated and over design to just test the newly added logic in this PR, and it introduces duplicate methods `setupMemory` and `newQueryCtx`.

So I also added another way to test  (**`RowNumberTest.reclaimFail`**), the 'hasSpilled' check logic added in this PR. It injected the test method into the production code and triggered the spill no matter whether ``RowNumber::maybeReserve` triggered or not. It does not need the `setupMemory`, `newQueryCtx` method.

What do you think, and which one do you prefer? 

cc @xiaoxmeng

mbasmanova commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -218,4 +275,40 @@ TEST_F(RowNumberTest, largeInput) {
+    bool expected = e.errorCode() == error_code::kMemCapExceeded.c_str() ||
+        e.errorCode() == error_code::kMemAborted.c_str() ||
+        e.errorCode() == error_code::kMemAllocError.c_str();
+    if (!expected) {
```
Use ASSERT_ macros instead.

mbasmanova reviewed (changes requested):
Thank you for the fix. Looks great. The test needs some work though.

mbasmanova commented on velox/exec/tests/SharedArbitratorTest.cpp:
```diff
@@ -3859,14 +3859,20 @@ TEST_F(SharedArbitrationTest, concurrentArbitration) {
           }
         } catch (const VeloxException& e) {
-          VELOX_CHECK(
+          bool expected =
```
Looks like this code has the same problem of not ensuring that the code in 'try' actually throws.

duanmeng commented:
@mbasmanova @xiaoxmeng Hi Masha and Xuan, I've added the RowNumber task back to the concurrent test, and updated this PR's title and description. Could you please help to take another look?

mbasmanova reviewed (approved):
Thanks.
